### PR TITLE
Enable end-to-end encryption

### PR DESCRIPTION
**Pull request by msujew**
Refactors the messaging mechanism to include e2e encryption. Also refactors messages to use binary encoding (msgpack) during transmission to enable sending around binary payloads.

### CODE REVIEW

File: packages/open-collaboration-vscode/src/collaboration-status-service.ts [hunk 1]
```diff
@@ -57,10 +62,10 @@ export class CollaborationStatusService {
                 this.statusBarItem.text = '$(live-share) Share';
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/open-collaboration-vscode/src/collaboration-status-service.ts]: I've adjusted the status bar text to make it more unique.

**Review by spoenemann — APPROVED** [review]
Looks good to me. Great contribution! 🚀

**Review by spoenemann — CHANGES_REQUESTED** [review]
I need to revoke my approval. OpenPGP.js is LGPL licensed, which could have severe consequences for us and everyone using the project. We should look for an alternative, or find a way to integrate more loosely with the library (avoid a direct dependency).

**Review by jonah-iden — APPROVED** [review]
Works great even with 3 users in one room and all in all looks good to me

File: packages/open-collaboration-rpc/src/connection.ts [hunk 2]
```diff
@@ -78,53 +81,94 @@ export class AbstractBroadcastConnection implements BroadcastConnection {
...
-                    const result = handler(message.origin, ...(message.params ?? []));
-                    Promise.resolve(result).then(value => {
-                        const responseMessage = ResponseMessage.create(message.id, value);
-                        this.write(responseMessage);
-                    }, error => {
-                        const responseErrorMessage = ResponseErrorMessage.create(message.id, error.message);
-                        this.write(responseErrorMessage);
+                    const result = handler(decrypted.origin, ...(decrypted.content.params ?? []));
+                    await this._ready.promise;
+                    Promise.resolve(result).then(async value => {
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/open-collaboration-rpc/src/connection.ts]: Yeah, in fact using the normal `try-catch` is actually way less code. Thanks!

File: packages/open-collaboration-rpc/src/encryption.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        const keyBuffer = fromBase64(key);
+        const content = message.content;
+        const encoded = Encoding.encode(content);
+        const compressionAlgo = Compression.bestFit(keys.map(key => key.supportedCompression));
+        const compressed = await Compression.compress(encoded, compressionAlgo);
+        const iv = await cryptoLib.generateIV();
+        const encrypted = await cryptoLib.symEncrypt(compressed, key, iv);
+        const encryptedKeys = await Promise.all(keys.map(async key => ({
+            target: key.peerId,
+            key: toBase64(await cryptoLib.publicEncrypt(keyBuffer, key.publicKey)),
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/open-collaboration-rpc/src/encryption.ts]: It's an implementation detail (i.e. doesn't affect the working of the protocol) but I've added the caching now.

**Comment by jonah-iden** [comment]
@msujew can confirm caching works great and it feels super fast

**Review by spoenemann — APPROVED** [review]
The license issue is resolved.

**Pull request merged**